RFC_STORE_MODIFIER.md

```diff
@@ -0,0 +1,185 @@
+
+If `replaceReducer` is called in `modifyStore`, this component wraps the children with a new context-provider with an updated `storeState`. The provider is only rendered on the first render pass since `storeState` will be correct after this.
+
+**Arguments**
+
+1. `modification` is an object that will be passed into `modifyStore`
+
+2. `modifyStore` is an optional prop that allows for passing in a `modifyStore`-function directly to `<StoreModifier>`. This makes it possible to avoid having to pass it down via context and simplifying the API when SSR-support is not needed (libraries could support both versions).
+
+3. `children` are the children that needs a modified store
+
+#### `withModifiedStore(modification, options)(WrappedComponent)`
+
+This is a Higher-Order-Component that wraps the `WrappedComponent` in a `<StoreModifier>`. It additionally takes care of hoisting statics on `WrappedComponent` and can optionally `forwardRef` to the wrapped component. As opposed to using `<StoreModifier>` directly, this HOC also supports using a custom `context` (if you are using something other than the default `ReactReduxContext`).
+
+**Arguments**
+
+1. `modification` is an object that will be passed into `modifyStore`
+
+2. `options` is an optional object with the shape of: `{ context, forwardRef, modifyStore }`. `context` can be used to consume a custom context. If `forwardRef` is true, adding a ref to the connected wrapper component will actually return the instance of the wrapped component, (both these correspond to the same options in `connect`). `modifyStore` is for passing in a custom `modifyStore`-function, see description in `<StoreModifier>`.
+
+3. `WrappedComponent` is the component to be wrapped with `<StoreModifier>`
+
+**Returns**
+
+A component that wraps the `WrappedComponent` in a `StoreModifier`
+
+#### `createStoreModifier(context)`
+
+This function returns a `StoreModifier`-component that uses the custom context provided to it. Can be used when a custom context is needed, but you don't want to use the HOC that supports this out of the box (libraries that want to avoid an extra component nesting for example).
+
+**Arguments**
+
+1. `context` is the custom context that you want the resulting component to use
+
+**Returns**
+
+A `StoreModifier`-component that uses the `context` passed in
+
+### Unsolved questions
+
+- Can naming be improved?
+- It would be possible to support a middleware-style API for `modifyStore`, [as per this comment](https://github.com/reduxjs/react-redux/issues/1150#issuecomment-450596584). This could be implemented in userland, but it might make sense to add this as the official API to encourage modular `storeModifiers`?
+  - In that case: How do we deal with namespacing-issues in the `modifier`-object?
```

Comment:
I'm leaning towards **Option 1**, but with the caveat of providing advice along these lines in the documentation (better formulated..):

> When using multiple modifiers, or if you are a library author, we recommend creating utility-functions for creating the `modification`-object, for example: `addReducersModification(newReducers)`. This should work as an action-creator and could look something like this:
>
> ```jsx
> const addReducersModification = newReducers => ({
>     type: '@your-library-name/type-of-modification',
>     payload: newReducers
> });
> ```
> 
> Then make sure your modifier-functions only respond to the modifications with the correct type.

The `modification`-object really is very similar to an action, just describes a modification supposed to happen to the store instead. For actions we handle namespacing via `type`, so why not reuse that concept here?

It would of course be possible to rename `modification` to `action` instead, but I think that might be more confusing than it clarifies things?